edx_rest_framework_extensions/auth/jwt/middleware.py

```diff
@@ -40,29 +41,45 @@ class EnsureJWTAuthSettingsMiddleware(MiddlewareMixin):
     """
     _required_permission_classes = (NotJwtRestrictedApplication,)
 
+    @staticmethod
+    def _get_decomposed_permissions(permission):
+        """
+        Given a composed permission class or instance, return its component parts.
+
+        Examples:
+
+        Composition using DRF native support in 3.9+:
+            IsStaff | IsSuperuser -> [IsStaff, IsSuperuser]
+            IsOwner | IsStaff | IsSuperuser -> [IsOwner | IsStaff, IsSuperuser]
+
+        """
+        # We skip SingleOperandHolder since it contains negations
```

Comment:
I wonder if it really is ok to skip negations. If so, maybe we need to better document why.
1. Can the negation be of a composed permission? Can there be double negations?
2. Since we aren't really checking for sound boolean logic, but just making sure a particular permission is somehow included, I wonder if we could inadvertently be causing issues by skipping negations. Or, does this end up matching earlier logic? Do we have a test around this from before/after?

Reply:
The reason I'm not including negation is that this middleware is about making sure that the `NotJwtRestrictedApplication` permission is required for all view classes that use JWT. If the `NotJwtRestrictedApplication` class is applied in a negation, then it is explicitly not applied since the permission is checking for the absence of that permission. 

I think, taking a step back and looking at the reason for this middleware, though, it might be OK to count negations. The reason being that any place that is explicitly negating this permission while using JWT is already aware of this permission and is explicitly crating a set of permissions based on that.

As for inadvertently causing issues. rest_condition and composed permissions in general are used very sparingly in the platform, and I have a [PR ](https://github.com/edx/edx-platform/pull/28663) that touches that code so we can see already that such a case isn't currently present in the platform. 

It may affect plugins, but then the move from rest condition to DRF will be breaking in that case anyway.